### ISSUE TITLE
RQ worker fails to update stats (MySQL server has gone away)

### ISSUE THREAD

**Issue by gwaldvogel** [translate/pootle#4692]
Every time the RQ worker tries to update the stats it fails saying "MySQL server has gone away".
RQ worker log:
```
Traceback (most recent call last):
  File "/var/www/pootle/env/local/lib/python2.7/site-packages/rq/worker.py", line 568, in perform_job
    rv = job.perform()
  File "/var/www/pootle/env/local/lib/python2.7/site-packages/rq/job.py", line 495, in perform
    self._result = self.func(*self.args, **self.kwargs)
  File "/var/www/pootle/env/local/lib/python2.7/site-packages/pootle/core/mixins/treeitem.py", line 684, in update_cache_job
    instance._update_cache_job(keys, decrement)
  File "/var/www/pootle/env/local/lib/python2.7/site-packages/pootle/core/mixins/treeitem.py", line 529, in _update_cache_job
    self.update_cached(key)
  File "/var/www/pootle/env/local/lib/python2.7/site-packages/pootle/core/mixins/treeitem.py", line 49, in _statslog
    result = function(instance, *args, **kwargs)
  File "/var/www/pootle/env/local/lib/python2.7/site-packages/pootle/core/mixins/treeitem.py", line 307, in update_cached
    self.set_cached_value(name, self._calc(name))
  File "/var/www/pootle/env/local/lib/python2.7/site-packages/pootle/core/mixins/treeitem.py", line 207, in _calc
    return self._calc_last_action()
  File "/var/www/pootle/env/local/lib/python2.7/site-packages/pootle/core/mixins/treeitem.py", line 166, in _calc_last_action
    [self._get_last_action()] +
  File "/var/www/pootle/env/local/lib/python2.7/site-packages/pootle/apps/pootle_store/models.py", line 2354, in _get_last_action
    .exclude(type=SubmissionTypes.UNIT_CREATE) \
  File "/var/www/pootle/env/local/lib/python2.7/site-packages/pootle/apps/pootle_statistics/models.py", line 110, in latest
    return self._earliest_or_latest(field_name=field_name, direction="-")
  File "/var/www/pootle/env/local/lib/python2.7/site-packages/pootle/apps/pootle_statistics/models.py", line 104, in _earliest_or_latest
    return obj.get()
  File "/var/www/pootle/env/local/lib/python2.7/site-packages/django/db/models/query.py", line 351, in get
    num = len(clone)
  File "/var/www/pootle/env/local/lib/python2.7/site-packages/django/db/models/query.py", line 122, in __len__
    self._fetch_all()
  File "/var/www/pootle/env/local/lib/python2.7/site-packages/django/db/models/query.py", line 966, in _fetch_all
    self._result_cache = list(self.iterator())
  File "/var/www/pootle/env/local/lib/python2.7/site-packages/django/db/models/query.py", line 265, in iterator
    for row in compiler.results_iter():
  File "/var/www/pootle/env/local/lib/python2.7/site-packages/django/db/models/sql/compiler.py", line 701, in results_iter
    for rows in self.execute_sql(MULTI):
  File "/var/www/pootle/env/local/lib/python2.7/site-packages/django/db/models/sql/compiler.py", line 787, in execute_sql
    cursor.execute(sql, params)
  File "/var/www/pootle/env/local/lib/python2.7/site-packages/django/db/backends/utils.py", line 65, in execute
    return self.cursor.execute(sql, params)
  File "/var/www/pootle/env/local/lib/python2.7/site-packages/django/db/utils.py", line 94, in __exit__
    six.reraise(dj_exc_type, dj_exc_value, traceback)
  File "/var/www/pootle/env/local/lib/python2.7/site-packages/django/db/backends/utils.py", line 65, in execute
    return self.cursor.execute(sql, params)
  File "/var/www/pootle/env/local/lib/python2.7/site-packages/django/db/backends/mysql/base.py", line 129, in execute
    return self.cursor.execute(query, args)
  File "/var/www/pootle/env/local/lib/python2.7/site-packages/MySQLdb/cursors.py", line 205, in execute
    self.errorhandler(self, exc, value)
  File "/var/www/pootle/env/local/lib/python2.7/site-packages/MySQLdb/connections.py", line 36, in defaulterrorhandler
    raise errorclass, errorvalue
OperationalError: (2006, 'MySQL server has gone away')`

Corresponding MySQL log:
`[Warning] Aborted connection 69 to db: 'pootledb' user: 'pootle' host: 'localhost' (Got timeout reading communication packets)
```

My system config:
Ubuntu 14.10

```
$ pootle --version
Pootle 2.7.2 (Django 1.7.11, Translate Toolkit 1.13.0)
$ mysql --version
mysql  Ver 14.14 Distrib 5.6.19, for debian-linux-gnu (x86_64) using  EditLine wrapper
```

I have set CONN_MAX_AGE to 0 and adapted the MySQL config:
```
wait_timeout		= 30
interactive_timeout	= 30
max_allowed_packet	= 256M
[...]
innodb_log_buffer_size 	= 64M
innodb_log_file_size	= 2047M
```

I am running two RQ workers with the following upstart script:
```bash
description     "Pootle rqworker"
author          "Gregor Waldvogel"

start on started apache2
stop on stopping apache2

env POOTLE_SETTINGS=/var/www/pootle/pootle.conf
exec su -s /bin/sh -c 'exec "$0" "$@"' www-data -- /var/www/pootle/env/bin/python /var/www/pootle/env/bin/pootle rqworker

respawn
```

**Comment by ta2-1** [comment]
Hi @gwaldvogel, I think you hit this issue https://github.com/translate/pootle/issues/4094 which was already fixed. 
We use per request Django approach and call `close_if_unusable_or_obsolete()` before and after every RQ job (https://github.com/translate/pootle/pull/4156).

**Comment by phlax** [comment]
@ta2-1 do you know - is it fixed in the current stable release (2.7.3) ?

**Comment by ta2-1** [comment]
@phlax yes, it should be there.

**Comment by gwaldvogel** [comment]
Hi @ta2-1 that was my first thought too, but I was able to reproduce it in my dev environment (running Pootle version 2.7.3). I will update my production environment tomorrow and see if it still occurs with v2.7.3. It is possible that I messed up my dev environment and this Issue is already fully resolved by #4156

**Comment by gwaldvogel** [comment]
Nevermind, the problem no longer occurs on my production system after upgrading to 2.7.3.
